faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
+    # Most likely this leads to Nim collecting some object prematurely.
+    # h.s = nil
+    # We work-around the problem through more indirect incapacitatation
+    # of the stream object:
+    h.s.preventFurtherReading()
+
+converter implicitDeref*(h: InputStreamHandle): InputStream =
+  h.s
+
+let FileStreamVTable = InputStreamVTable(
+  closeSync: proc (s: InputStream)
+                  {.nimcall, gcsafe, raises: [IOError, Defect].} =
+    try:
+      close FileInputStream(s).file
+    except OSError as err:
+      raise newException(IOError, "Failed to close file", err)
+  ,
+  getLenSync: proc (s: InputStream): int
+                   {.nimcall, gcsafe, raises: [IOError, Defect].} =
+    distance(s.head, s.bufferEnd)
+)
+
+template vtableAddr*(vtable: InputStreamVTable): ptr InputStreamVTable =
+  ## This is a simple work-around for the somewhat broken side
+  ## effects analysis of Nim - reading from global let variables
+  ## is considered a side-effect.
+  {.noSideEffect.}:
+    unsafeAddr vtable
+
+proc fileInput*(filename: string): InputStreamHandle =
+  let
+    memFile = memfiles.open(filename)
+    head = cast[ptr byte](memFile.mem)
+    fileSize = memFile.size
+
+  var stream = FileInputStream(
+    vtable: vtableAddr FileStreamVTable,
+    head: head,
+    bufferEnd: offset(head, fileSize),
+    bufferEndPos: fileSize,
+    file: memFile)
 
-proc openFile*(filename: string): ByteStreamVar =
-  new result
-  var memFile = memfiles.open(filename)
-  result.head = cast[ptr byte](memFile.mem)
   when debugHelpers:
-    result.bufferStart = result.head
-  result.bufferEnd = offset(result.head, memFile.size)
-  result.bufferEndPos = memFile.size
-  result.closeStream = proc = close memFile
-
-proc init*(T: type ByteStream,
-           mem: openarray[byte],
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): ByteStream =
-  # TODO: the result should use `from mem` once it's supported
-  result.head = unsafeAddr mem[0]
-  result.bufferEnd = offset(result.head, mem.len)
-  result.bufferEndPos = mem.len
-  result.reader = reader
-  result.asyncReader = asyncReader
-
-proc memoryStream*(mem: openarray[byte]): ByteStreamVar = # TODO: mark the result with `from mem`
-  new result
-  result[] = ByteStream.init(mem)
-
-proc memoryStream*(str: string): ByteStreamVar = # TODO: mark the result with `from str`
-  new result
-  result[] = init(ByteStream, str.toOpenArrayByte(0, str.len - 1))
-
-proc init*(T: type BufferedStream,
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): BufferedStream =
-  result.init result.buffer, reader, asyncReader
-
-proc endPos*(s: InputStreamVar): int =
-  # TODO This needs to use a VTable for `system.File` based streams
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
```

Comment:
Well, the streaming API is one of the primary reasons (but not the only one) for why I consider https://github.com/nim-lang/RFCs/issues/178 to be the most important improvement of Nim for the purposes of our project. With the lifetime tracking envisioned there, the compiler will be able to ensure that the backing object outlives the created memory stream. 

In general, I simply think that we cannot have designs that are friendly to resource-restricted devices without employing streaming and we cannot afford to build everything with APIs that require you to produce a copied `seq[byte]` every time you need to extract a slice of a network message.

So, the solution for me is to require for now that the usages of `memoryInput` are carefully reviewed. Luckily, `memoryInputs` are used very rarely in practice. In our current codebase, they are used only in tests and in functions such as `Json/SSZ.decode` which serve as safe examples.

Reply:
well, the issue is kind of exactly the same with files - and worse so with memory mapped files - when you close a file descriptor, the worst that happens with references to it is that the next read returns a predictable error that you can usually handle (until the descriptor is recycled I guess) - when you close a memfile or let a memory reference go out of scope, you get a hard crash or a security issue.

basically, the order in which to implement things here is to make 178 pass first and then start relying on it in mission-critical code - otherwise we're asking our users to take the risk for us.

> network message

our network messages are small - we mostly don't need to stream them, really - they're already chunked into manageable pieces. we can wait for the right language features before this actually starts to matter (if ever) - not crashing is more important to us from a purely pragmatic point of view.

Reply:
It's easy to create a `safeMemoryInput` option that first copies the data into a buffer managed by the stream, but this seems a bit overly conservative to me. In practice, it's easy to review the usages and you'll see that the `memoryInput` objects are so short-lived that it's trivial to reason about their safety:

https://github.com/status-im/nim-serialization/blob/master/serialization.nim#L54-L56

Reply:
there's only so many "special" cases you can remember during a review - or indeed when coding things - we already pretty much decided to not allow untracked references to leak through the ranges library because the risk is greater than the benefit. 

I'd be happier to accept this if we had a common name for anything that does unsafe stuff - the list is growing - in the language, that's `addr`, `unsafeddr`, `ptr` and `pointer` basically (did I get them all?), but now it's proliferating and having this discussion about "is my special case special enough to be special" over and over saps developer energy. 

why don't we call this one `unsafeMemoryInput` to rhyme with `unsafeAddr` and we'll be done with it - it's ugly but that's the point - then we can measure how often it happens and either motivate a new language feature based on actual data or use the safe variant where it doesn't matter - ie reading a json rpc command from a socket, I couldn't care less if there's a memory copy involved but I do care that the compiler maintains a safety net for me until my feature is solid and copying the buffer is my biggest issue - at that point, I'll gladly type out the extra 6 characters.

Reply:
I like the `unsafeMemoryInput` idea. I'll rename it.